refactor: createRouter + NavLink 클래스 추가

jihyeon-songb opened the pull request:
- 전역 이벤트 리스너의 과도한 책임 축소를 위해 전역 [data-link] 클릭 핸들러 제거 → 링크별 NavLink 인스턴스에서 라우팅 처리 ([ef75359](https://github.com/f-lab-edu/my-spa-router/pull/2/commits/ef753599243b6e27b38026142da636fd67269d14))
- 확장성 확보를 위해 기존 router() 함수는 제거하고 createRouter로 전환 ([ef75359](https://github.com/f-lab-edu/my-spa-router/pull/2/commits/ef753599243b6e27b38026142da636fd67269d14))

jihyeon-songb committed:
feat: SPA 라우터 서버 및 HTML 초기 세팅

jihyeon-songb committed:
feat: src/pages → src/views로 폴더 구조 변경 및 내부 코드 추가

jihyeon-songb committed:
feat: SPA 라우터 기본 기능 구현

pro-flab commented on front/src/main.js:
```diff
@@ -0,0 +1,38 @@
+import Posts from './views/Posts.js';
+import Settings from './views/Settings.js';
+
+document.body.addEventListener('click', async e => {
+    if (e.target.matches('[data-link]')) {
+        // 동작 멈춤
+        e.preventDefault();
+        let path = e.target.href;
+        history.pushState(null, null, path);
+        await router();
+    }
+})
```
여러 가지 문제가 있습니다.

- 전역 이벤트 리스너가 너무 과도한 책임을 가지고 있습니다. 이 클릭 이벤트 리스너가 모든 라우팅 이동을 담당하고 있어요.
- 이로 인해 유연성이 부족해집니다. 모든 [data-link] 요소가 동일한 방식으로 처리되는데, 다른 써드파티 라이브러리나 다른 개발자들이 [data-link]인 요소를 쓰지 않을거라는 가정하에 작성되었는데 이것은 좋지 않은 방향입니다.
- 전역 이벤트 델리게이션으로 인해 특정 링크의 동작을 추적하기 어려워서 디버깅이 매우 어려워질겁니다.

차라리 아래와 같이 네비게이션을 가능하게 하는 링크를 만드는 로직을 만들면 어떨까 합니다.

```
// NavLink 컴포넌트
class NavLink {
  constructor(element) {
    this.element = element;
    this.href = element.dataset.href;
    this.beforeNavigate = element.dataset.beforeNavigate;
    
    this.element.addEventListener('click', this.handleClick.bind(this));
  }

  async handleClick(e) {
    e.preventDefault();
    
    // 네비게이션 전 훅
    if (this.beforeNavigate && !await this.runBeforeNavigate()) {
      return;
    }
    
    router.navigate(this.href);
  }

  async runBeforeNavigate() {
    // 조건부 네비게이션 로직
    return true;
  }
}
```

pro-flab commented on front/src/main.js:
```diff
@@ -0,0 +1,38 @@
+
+    const html = await match.view.render();
+    app.innerHTML = html;
+}
```
요기도 문제점이 있습니다.

1. popstate가 일어날 때마다 매번 router를 호출함
  - routes 배열이 매번 새로 생성됨
  - 라우트 매칭 로직이 불필요하게 반복 실행됨

2. 근본적으로 라우터 관련 로직이 책임이 분리되어 있고 router 함수로 퉁치고 있어서 발생함.
  - 초기화와 라우트 처리가 하나의 함수에 섞여 있음
  - 라우트 등록/초기화과 라우트 실행 로직이 분리되지 않음

3. 확장성 문제
  - 새로운 라우트 추가 시 router 함수 내부를 수정해야 함

이런 관점에서 AI의 도움을 받아 개선된 코드를 작성하고, 어떻게 개선되었는지 설명해주세요!

pro-flab commented on server.js:
```diff
@@ -0,0 +1,12 @@
+const express = require('express');
+const path = require('path');
+const app = express();
+const PORT = process.env.PORT || 3000;
+
+app.use(express.static(path.join(__dirname, 'front')));
+
+app.get('/*', (req, res) => {
+    res.sendFile(path.join(__dirname, 'front', 'index.html'));
+});
+
+app.listen(PORT, () => console.log(`http://localhost:${PORT}`));
```
서버를 따로 구현해야 했었던 이유가 있나요??

jihyeon-songb committed:
refactor: createRouter + NavLink 클래스 추가

- 전역 이벤트 리스너의 과도한 책임 축소를 위해 전역 [data-link] 클릭 핸들러 제거 → 링크별 NavLink 인스턴스에서 라우팅 처리
- 확장성 확보를 위해 기존 router() 함수는 제거하고 createRouter로 전환

pro-flab commented on front/src/main.js:
```diff
@@ -0,0 +1,17 @@
+
+document.addEventListener('DOMContentLoaded', () => {
+    document.querySelectorAll('#main-nav a[href]').forEach(el => new NavLink(el, router));
+});
```
음.. 매번 querySelectorAll을 잡아서 **암묵적으로** NavLink를 생성해주는건 좋지 않은 것 같아요. 필요한 곳에서 NavLink를 쓰게 하는게 좋을 것 같아요.

여기서 중요한 포인트는 이 동작이 **전역적으로 암묵적**으로 동작한다는 사실인데요, 이것이 무얼 의미하는지 LLM과 함께 학습해주셨으면 좋겠어요.

필요한 곳에서 NavLink를 사용해서 쓰도록 한 번 바꿔볼까요?

pro-flab replied:
그리고 근본적으로 DOMContentLoaded는 컨텐츠가 로드된 이후에 DOM이 변경되었을 경우에는 호출되지 않아요. 이후 자바스크립트로 인해 DOM 변경이 일어난다면 동작하지 않을거예요.

pro-flab commented on front/src/createRouter.js:
```diff
@@ -0,0 +1,26 @@
+        await render(path);
+    };
+
+    const setRoutes = (r) => { routes = r; };
```
setRoutes는 사용하지 않는 + 제대로 동작하지 않는 함수인 것 같아요. 이렇게 사용하지도 않고 동작하지 않는 함수를 함부러 노출했다가 이 라이브러리를 사용하는 사람이 만약 사용했다면 엄청난 문제가 발생할 수 있겠죠! 삭제해주시면 좋을 것 같아요

pro-flab commented on front/src/createRouter.js:
```diff
@@ -0,0 +1,26 @@
+    window.addEventListener('popstate', () => render(location.pathname));
+    document.addEventListener('DOMContentLoaded', () => render(location.pathname));
+
+    return { navigate, render, resolve, setRoutes };
```
`render`와 `resolve` 함수 또한 라우터 내부적으로 사용하는 함수여야만 하는 것 같아요. 라우터 외부에서 누군가가 render, resolve 같은 함수를 오용하게 된다면 문제가 발생할 수 있는 구조 같아요. 이런건 노출하지 않아야 해요.

pro-flab commented on front/src/createRouter.js:
```diff
@@ -0,0 +1,26 @@
+
+    // 브라우저 이동 처리
+    window.addEventListener('popstate', () => render(location.pathname));
+    document.addEventListener('DOMContentLoaded', () => render(location.pathname));
```
1. 둘 다 `document.addEventListener`로 해주시면 좋을 것 같습니다.
2. 이벤트 리스너는 주의해야 할 점이 있어요. 바로 이벤트 리스너를 등록한 후 명시적으로 해제해주지 않으면 메모리 누수가 일어난다는 것인데요. 이에 대해 자세히 공부해봐주시고 이를 해결하기 위해서 어떻게 해야할지 ChatGPT 등을 이용해 학습해서 개선해주세요.

jihyeon-songb committed:
refactor: Router 클래스 도입 및 NavLink 바인딩 명시화

- createRouter.js 삭제, Router 클래스로 교체
- routes 구조 정리
- render를 프라이빗 메서드(#render)로 전환
- 암묵적 NavLink 생성 제거, 필요한 요소에만 명시적으로 바인딩

pro-flab commented on front/src/main.js:
```diff
@@ -0,0 +1,25 @@
+];
+
+export const router = new Router({ outlet: '#app', routes });
+for (const route of routes) {
+    new NavLink(document.getElementById(route.id), router);
+}
```
전체적으로 제가 의도한 방향과 크게 다르지는 않는데요.
NavLink 또한 페이지 자체에 포함되면 되지 않을까요?

저는 이런 형태의 인터페이스를 원했어요. 실제로 React에서 react-router의 `Link` 컴포넌트를 쓰는 것 처럼요.

```tsx
// Home.js
const Home = {
    async render() {
        return `<h1>HOME</h1>{NavLink.to('/posts')}`;
    }
}
export default Home;
```

pro-flab commented on front/src/components/Router.js:
JSDoc으로 타입 정의를 넣어줘

pro-flab commented on front/src/views/Settings.js:
```diff
@@ -0,0 +1,6 @@
+const Settings = {
+    async render() {
+        return `<h1>Settings</h1>`;
```
NavLink를 이미 HTML 트리에 있는 anchor를 순회하면서 그 anchor에 기능을 집어넣는 것이 아니라, 컴포넌트 렌더링 과정에 추가할 수 있도록 구현해봐주세요!
